plane/src/typed_unix_socket/client.rs

```diff
@@ -0,0 +1,247 @@
+    shutdown_rx: watch::Receiver<()>,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync>>
+where
+    MessageToServer: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+    MessageToClient: Send + Sync + 'static + Clone + Debug + Serialize + for<'de> Deserialize<'de>,
+{
+    let (read_half, write_half) = tokio::io::split(stream);
+    let reader = BufReader::new(read_half);
+    let writer = BufWriter::new(write_half);
+
+    let mut lines = reader.lines();
+    let mut writer = writer;
+
+    // Task to handle receiving messages
+    let recv_task = tokio::spawn({
+        let event_tx = event_tx.clone();
+        let response_map = Arc::clone(&response_map);
+        let mut shutdown_rx = shutdown_rx.clone();
+        async move {
+            loop {
+                tokio::select! {
+                    _ = shutdown_rx.changed() => {
+                        tracing::info!("Shutting down receive task");
+                        break;
+                    }
+                    result = lines.next_line() => {
+                        match result {
+                            Ok(Some(line)) => {
+                                let msg: WrappedMessage<MessageToClient> = match serde_json::from_str(&line)
+                                {
+                                    Ok(msg) => msg,
+                                    Err(e) => {
+                                        tracing::error!("Error deserializing message: {}", e);
+                                        continue;
+                                    }
+                                };
+                                match msg {
+                                    WrappedMessage {
+                                        id: Some(id),
+                                        message,
+                                    } => {
+                                        if let Some((_, tx)) = response_map.remove(&id) {
+                                            if let Err(e) = tx.send(message) {
+                                                tracing::error!("Error sending response: {:?}", e);
+                                            }
+                                        } else {
+                                            tracing::error!("No sender found for response ID: {:?}", id);
+                                        }
+                                    }
+                                    WrappedMessage { id: None, message } => {
+                                        if let Err(e) = event_tx.send(message) {
+                                            tracing::error!("Error sending event: {}", e);
+                                        }
+                                    }
+                                }
+                            }
+                            Ok(None) => {
+                                tracing::error!("Connection closed by server");
```

Comment:
Should this `break`?

Reply:
+1

Also, if you break here, I don't think you need the whole `shutdown` part, it seems like you are just reimplementing the connection closure mechanism that already exists on a channel.